### DISCUSSION TITLE
OffSets and rotations in parallel geometry

### DISCUSSION THREAD

**Discussion by Vilax** [CERN/TIGRE#572] [Q&A]
Hi Ander,

I am impressed with TIGRE, really nice job, thanks for gift and congratulations

The question I have is the following one: 

- The source and the detector are static, only the sample is rotated.
- The sample is tilted at different angles, resulting in a set of images.
- The tilting process introduces misalignment: Rotations around the beam direction (I assume is X in your geometry) and shifts in the perpendicular plane to the beam direction (I assume YZ in your geometry)
- So per image, I know the transformation matrix to align the image.  This means to make that all images share the same tilt axis. I can align the image before enter into TIGRE, but I prefer to do it with TIGRE to avoid interpolation errors

The problem I have is: I have only shifts I can correct them. If I have only rotations, I can correct them. But if I have shifts and rotations... The reconstruction is a disaster. I paste a draft of my code just in case you can cast light on me

xdim, ydim, thickness = 1000, 1000, 300

for image in setofimages:
    offSets.append([sy, sx])               # each image has a shift in the perpendicular plane to the beam
    rotAngles.append([-rot, 0.0, 0.0]) # each image has a rotation in XZ
    tiltAngles.append([0.0, 0.0, tilt])   # tilt angle of the sample

tiltAngles = np.vstack(tiltAngles) * np.pi/180.0
rotAngles  = np.vstack(rotAngles) * np.pi/180.0
offSets    = np.vstack(offSets)

geo = tigre.geometry(mode="parallel", nVoxel=np.array([xdim, ydim, thickness]))
geo.rotDetector = rotAngles
geo.offDetector = offSets

reconstruction = algs.fbp(ts, geo, tiltAngles, filter=self.filterToApply, noneg=False, gpuids=gpuids)

Thanks in advance

NOTE: This is similar, but  different to the topic https://github.com/CERN/TIGRE/discussions/565

**Comment by AnderBiguri** [comment]
so are you doing some sort of laminography? if you axis of rotation is tilted you want to use the arbitrary axis of rotation demo. The way to define this is via 3vector rotational angles. 

rotDetector rotates teh detector _in place_, i,e, the plannar detector is moved w.r.t its center. I don't think this is what you are describing, as you say the detector is static.

**Reply by Vilax** [reply]
Thanks for your quick reply.

It is not exactly laminography, in this case is is electron tomography. The detector and the source are static in an electron microscope, only the sample is tilted.

I see in the arbitrary axis of rotation demo that you matrix follows a ZYZ convention. In our case, we tilt around Z (in agreement with your model) and the misalignment rotations of the sample are around X (in disagreement). So, I think a conversion of the Euler angles from ZXZ to ZYZ should fix the problem.

I will check, and keep you updated.

**Reply by Vilax** [reply]
Thanks, Ander,

We could fix the geometry issue with the change of reference system. To help other people that can have the same question, here is a small code

`for image in setofimages:
offSets.append([-sy, -sx]) # each image has a shift in the perpendicular plane to the beam

zyz_angles = [tilt, rot, 0.0]  # replace with actual values
r_zyz = R.from_euler('ZYZ', zyz_angles, degrees=True)
zxz_angles = r_zyz.as_euler('ZXZ', degrees=True)
tiltAngles.append(zxz_angles)


tiltAngles = np.vstack(tiltAngles) * np.pi/180.0
rotAngles = np.vstack(rotAngles) * np.pi/180.0
offSets = np.vstack(offSets)

geo = tigre.geometry(mode="parallel", nVoxel=np.array([xdim, ydim, thickness]))
geo.rotDetector = rotAngles
geo.offDetector = offSets

reconstruction = algs.fbp(ts, geo, tiltAngles, filter=self.filterToApply, noneg=False, gpuids=gpuids)`

Thanks for your help

**Reply by Vilax** [reply]
On the other hand, I also would like to know your opinion, if you agree, about the next scenario. The following images are 3D reconstructions using FBP, SIRT (20 iters) and SART (20 iters). The input data is identical in all cases. I would like to understand the stripes of SIRT, and the non sense result of SART, but I have no clues about what can be happening.

FBP: reconstruction = algs.fbp(ts, geo, tiltAngles, filter=self.filterToApply, noneg=False, gpuids=gpuids)
SIRT reconstruction = algs.sirt(ts, geo, tiltAngles, 20, lmbda=1.0, lmbda_red=0.999, verbose=True, noneg=False, gpuids=gpuids)
SART reconstruction = algs.fbp(ts, geo, tiltAngles, 20, lmbda=1.0, lmbda_red=0.999, verbose=True, noneg=False, gpuids=gpuids)

![image](https://github.com/user-attachments/assets/41a7281a-c57f-4ab8-a3ba-be29906dd2cc)

Sorry for asking so much

Thanks in advance

**Reply by AnderBiguri** [reply]
Hi @Vilax! Interesting results. I have no much experience with electron tomography (albeit I have colleages at the LMB that may be able to look at it) so not sure if I can help much. 
For SIRT: 20 iterations is very little (try more like 200), but indeed you are getting very weird artifacts. I wonder if the pre-processing done to the tomographs is compatible with iterative algorithms. When tomogarphs are pre-processed for FBP, often they are e.g. padded with a repeating value boundary condition, or just zeros, for example. This is very ditruptive to iterative algorithms, but not FBP, as the latter uses a high-pass filter, so this "flat" padding dissapears, while iterative algorithms assume the raw data is correct, which then destroys data-consistency checks if padded with constant values. This is just an example, but I know that ET does quite a lot of pre-processing, so it could be some of that. 

For SART: I have no idea, I've never seen something go this bad.I'd expected a better image than SIRT for sure. Must have to do with something in the data that makes SART very confused, but I genuinelly never seen something this non sense created by one of the algorithms...
